// ngx-skltn/esm2015/lib/services/skltn.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable, Inject } from '@angular/core';
import { SKLTN_CONFIG_TOKEN } from './skltn-config.service';
import * as i0 from "@angular/core";
import * as i1 from "./skltn-config.service";
export class SkltnService {
    /**
     * @param {?} config
     */
    constructor(config) {
        this.ids = [];
        this.defaultConfig = {
            rectRadius: 4,
            bgFill: '#ddd',
            flareFill: 'rgba(255, 255, 255, 0.6)',
            flareWidth: '150px',
            duration: 1200,
            delay: 0,
            timing: 'ease-in-out',
        };
        this.config = Object.assign({}, this.defaultConfig, config);
    }
}
SkltnService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] }
];
/** @nocollapse */
SkltnService.ctorParameters = () => [
    { type: undefined, decorators: [{ type: Inject, args: [SKLTN_CONFIG_TOKEN,] }] }
];
/** @nocollapse */ SkltnService.ngInjectableDef = i0.defineInjectable({ factory: function SkltnService_Factory() { return new SkltnService(i0.inject(i1.SKLTN_CONFIG_TOKEN)); }, token: SkltnService, providedIn: "root" });
if (false) {
    /** @type {?} */
    SkltnService.prototype.ids;
    /** @type {?} */
    SkltnService.prototype.config;
    /**
     * @type {?}
     * @private
     */
    SkltnService.prototype.defaultConfig;
}
