// ngx-skltn/esm5/lib/ngx-skltn.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { SkltnComponent } from './skltn/skltn.component';
import { SkltnBoneDirective } from './directives/skltn-bone.directive';
import { SkltnService } from './services/skltn.service';
import { SKLTN_CONFIG_TOKEN } from './services/skltn-config.service';
var NgxSkltnModule = /** @class */ (function () {
    function NgxSkltnModule() {
    }
    /**
     * @param {?=} config
     * @return {?}
     */
    NgxSkltnModule.forRoot = /**
     * @param {?=} config
     * @return {?}
     */
    function (config) {
        if (config === void 0) { config = {}; }
        return {
            ngModule: NgxSkltnModule,
            providers: [{
                    provide: SKLTN_CONFIG_TOKEN,
                    useValue: config,
                },
                {
                    provide: SkltnService,
                    useClass: SkltnService,
                    deps: [SKLTN_CONFIG_TOKEN],
                }]
        };
    };
    NgxSkltnModule.decorators = [
        { type: NgModule, args: [{
                    declarations: [
                        SkltnComponent,
                        SkltnBoneDirective,
                    ],
                    imports: [
                        CommonModule,
                    ],
                    exports: [
                        SkltnComponent,
                        SkltnBoneDirective,
                    ]
                },] }
    ];
    return NgxSkltnModule;
}());
export { NgxSkltnModule };
