// ngx-skltn/esm2015/lib/directives/skltn-bone.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, ElementRef, Input, ContentChild, TemplateRef } from '@angular/core';
export class SkltnBoneDirective {
    /**
     * @param {?} element
     */
    constructor(element) {
        this.element = element;
    }
}
SkltnBoneDirective.decorators = [
    { type: Directive, args: [{
                selector: '[skltn-bone]',
            },] }
];
/** @nocollapse */
SkltnBoneDirective.ctorParameters = () => [
    { type: ElementRef }
];
SkltnBoneDirective.propDecorators = {
    type: [{ type: Input }],
    rectRadius: [{ type: Input }],
    pathWidth: [{ type: Input }],
    pathHeight: [{ type: Input }],
    template: [{ type: ContentChild, args: ['boneTemp',] }]
};
if (false) {
    /** @type {?} */
    SkltnBoneDirective.prototype.type;
    /** @type {?} */
    SkltnBoneDirective.prototype.rectRadius;
    /** @type {?} */
    SkltnBoneDirective.prototype.pathWidth;
    /** @type {?} */
    SkltnBoneDirective.prototype.pathHeight;
    /** @type {?} */
    SkltnBoneDirective.prototype.template;
    /** @type {?} */
    SkltnBoneDirective.prototype.element;
}
