// ngx-skltn/esm2015/lib/ngx-skltn.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { SkltnComponent } from './skltn/skltn.component';
import { SkltnBoneDirective } from './directives/skltn-bone.directive';
import { SkltnService } from './services/skltn.service';
import { SKLTN_CONFIG_TOKEN } from './services/skltn-config.service';
export class NgxSkltnModule {
    /**
     * @param {?=} config
     * @return {?}
     */
    static forRoot(config = {}) {
        return {
            ngModule: NgxSkltnModule,
            providers: [{
                    provide: SKLTN_CONFIG_TOKEN,
                    useValue: config,
                },
                {
                    provide: SkltnService,
                    useClass: SkltnService,
                    deps: [SKLTN_CONFIG_TOKEN],
                }]
        };
    }
}
NgxSkltnModule.decorators = [
    { type: NgModule, args: [{
                declarations: [
                    SkltnComponent,
                    SkltnBoneDirective,
                ],
                imports: [
                    CommonModule,
                ],
                exports: [
                    SkltnComponent,
                    SkltnBoneDirective,
                ]
            },] }
];
